### PR TITLE
Error handling cleanup in solvers/flattening 5

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

File: src/solvers/flattening/boolbv_unary_minus.cpp [hunk 1]
```diff
@@ -23,66 +25,53 @@ bvt boolbvt::convert_unary_minus(const unary_exprt &expr)
...
 
-  const bvt &op_bv=convert_bv(op0);
+  const bvt &op_bv = convert_bv(op, width);
 
   bvtypet bvtype=get_bvtype(type);
-  bvtypet op_bvtype=get_bvtype(op0.type());
-  std::size_t op_width=op_bv.size();
+  bvtypet op_bvtype = get_bvtype(op.type());
 
   bool no_overflow=(expr.id()=="no-overflow-unary-minus");
```
Comment (danpoe) [inline src/solvers/flattening/boolbv_unary_minus.cpp]: `"no-overflow-unary-minus"` does not seem to be assigned anywhere. Is this still needed?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: baec7a17b0c862211bc32e0fe8836eb06d8e34b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84576476

File: src/solvers/flattening/boolbv_unary_minus.cpp [hunk 2]
```diff
@@ -96,7 +85,7 @@ bvt boolbvt::convert_unary_minus(const unary_exprt &expr)
   }
   else if(bvtype==bvtypet::IS_FLOAT && op_bvtype==bvtypet::IS_FLOAT)
   {
-    assert(!no_overflow);
+    INVARIANT(!no_overflow, "");
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_unary_minus.cpp]: missing message

File: src/solvers/flattening/boolbv_unary_minus.cpp (hunk 1 above)
Comment (danpoe) [reply src/solvers/flattening/boolbv_unary_minus.cpp]: @peterschrammel Do you know if we can remove "no-overflow-unary-minus"?

File: src/solvers/flattening/boolbv_unary_minus.cpp [hunk 3]
```diff
@@ -23,66 +25,53 @@ bvt boolbvt::convert_unary_minus(const unary_exprt &expr)
...
 
-      for(std::size_t j=0; j<tmp_op.size(); j++)
-      {
-        assert(i*sub_width+j<op_bv.size());
-        tmp_op[j]=op_bv[i*sub_width+j];
-      }
-
-      bvt tmp_result;
+      const auto sub_it = std::next(op_bv.begin(), sub_idx);
+      std::copy_n(sub_it, sub_width, std::back_inserter(tmp_op));
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_unary_minus.cpp:59]: I'm sure this is a good idea, but seems unrelated to the goal of the commit?

Comment (danpoe) [reply src/solvers/flattening/boolbv_unary_minus.cpp:59]: Ok, this has become a separate commit now.

**Review by tautschnig — APPROVED** [review]

**Review by tautschnig — CHANGES_REQUESTED** [review]
Failing unit test:
`java_bytecode/ci_lazy_methods/lazy_load_lambdas.cpp:17` with
```
Invariant check failed
File: std_types.h:1319 function: validate_type
Condition: signed bitvector type must have non-zero width
Reason: type.get_width() > 0
```

**Comment by danpoe** [comment]
@tautschnig I've removed those additional checks `type.get_width() > 0` now. Turns out zero-width bitvector types can legally appear due to e.g. 0-width members of bitfields.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7d629c5ae25910d1274f476b1327e84c43b73069).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86032809
